pkg/web/apps/apps_psql_repository.go

```diff
@@ -117,7 +119,131 @@ func (a *appsPostgreSQLRepository) GetAppByID(ID string) (*models.App, error) {
+	device := models.Device{}
+
+	sqlStatement := `
+		SELECT d.id, d.version_id, d.app_id, d.device_id, d.device_type, d.device_version
+		FROM device as d
+		WHERE d.device_id = $1 AND d.app_id = $2;`
+
+	if err := a.db.QueryRow(sqlStatement, deviceID, appID).
+		Scan(&device.ID, &device.VersionID, &device.AppID, &device.DeviceID, &device.DeviceType, &device.DeviceVersion); err != nil {
 
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &device, nil
+}
+
+// GetDeviceByVersionAndAppID returns a device by its version number and app ID
+func (a *appsPostgreSQLRepository) GetDeviceByVersionAndAppID(version string, appID string) (*models.Device, error) {
+	device := models.Device{}
+
+	sqlStatement := `
+		SELECT d.id, d.version_id, d.app_id, d.device_id, d.device_type, d.device_version
+		FROM device as d
+		WHERE d.app_id = $1 AND d.device_version = $2;`
+
+	if err := a.db.QueryRow(sqlStatement, appID, version).
+		Scan(&device.ID, &device.VersionID, &device.AppID, &device.DeviceID, &device.DeviceType, &device.DeviceVersion); err != nil {
+
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &device, nil
+}
+
+// GetAppByID retrieves an app by id from the database
+func (a *appsPostgreSQLRepository) GetAppByAppID(appID string) (*models.App, error) {
+	app := models.App{}
+
+	sqlStatment := `SELECT id,app_id,app_name FROM app WHERE app_id=$1 AND deleted_at IS NULL;`
+	err := a.db.QueryRow(sqlStatment, appID).Scan(&app.ID, &app.AppID, &app.AppName)
+
+	if err != nil {
+		log.Error(err)
+		if err == sql.ErrNoRows {
+			return nil, models.ErrNotFound
+		}
+		return nil, models.ErrInternalServerError
+	}
+
+	return &app, nil
+
+}
+
+// InsertVersionOrUpdateNumOfAppLaunches creates a new version row
+// or increments the num_of_app_launches counter if the version already exists
+func (a *appsPostgreSQLRepository) InsertVersionOrUpdateNumOfAppLaunches(version *models.Version) error {
+	sqlStatement := `
+		INSERT INTO version(id, version, app_id, disabled, disabled_message, num_of_app_launches, last_launched_at)
+		VALUES($1, $2, $3, $4, $5, $6, NOW())
+		ON CONFLICT (id)
+		DO UPDATE
+		SET num_of_app_launches = $6,
+		last_launched_at = NOW();`
+
+	_, err := a.db.Exec(sqlStatement, version.ID, version.Version, version.AppID, version.Disabled, version.DisabledMessage, version.NumOfAppLaunches)
+
+	if err != nil {
+		log.Error(err)
+		return models.ErrDatabaseError
+	}
+
+	return nil
+}
+
+// InsertDeviceOrUpdateVersionID creates a new device row in the device table
+func (a *appsPostgreSQLRepository) InsertDeviceOrUpdateVersionID(device models.Device) error {
+	sqlStatement := `
+		INSERT INTO device(id,version_id,app_id,device_id,device_type,device_version)
+		VALUES($1, $2, $3, $4, $5, $6)
+		ON CONFLICT (id)
```

Comment:
@craicoverflow How it will works if one device has more than one app/version installed? 
I'd like to suggest you do the implementation in the service and not in the repo layer as the following for example.

1. Check if has a row with the device and version 
2. If not create/insert
3. If yes, them update the totals in the version table

c/c @davidffrench

Reply:
Check my suggestion https://github.com/aerogear/mobile-security-service/pull/76#discussion_r which also solved this scenario.